### PR TITLE
libtorch-bin: 2.0.0 -> 2.2.2

### LINKED ISSUE #299311: Update request: libtorch-bin 2.0.0 → 2.2.1
- Package name: libtorch-bin
- Latest released version: 2.2.1

- Current version on the unstable channel: 2.0.0
- Current version on the stable/release channel: 2.0.0

- [X] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@junjihashimoto 

-----

Note for maintainers: Please tag this issue in your PR.

libtorch 2.2.1 has a version for aarch64-darwin-cpu, while the current version on the unstable channel does not. I am specifically after Apple silicon mac support, so it's important for me that the aarch64-darwin version is added.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by pinpox**
## Description of changes

Also fixed the prefetch script, the url seems to have slightly changed Fixes #299311

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by msanft** [comment]
Result of `nixpkgs-review pr 302427` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>ocamlPackages.torch</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>in-formant</li>
    <li>libtorch-bin</li>
    <li>libtorch-bin.dev</li>
  </ul>
</details>

**Comment by pinpox** [comment]
> [](https://github.com/Mic92/nixpkgs-review)
> 1 package failed to build:
> 
>     * ocamlPackages.torch

Hm, not sure wahat to do about the ocaml package. @bcdarwin could you take a look?

**Review by junjihashimoto — CHANGES_REQUESTED** [review]
@pinpox Thank you for updating it.
Could you change cu118 to cu121? Because pytorch's default(pip) is cu121.

**Comment by pinpox** [comment]
> Could you change cu118 to cu121? Because pytorch's default(pip) is cu121.

done.

**Comment by bcdarwin** [comment]
> > 1 package failed to build:
> > ```
> > * ocamlPackages.torch
> > ```
> 
> Hm, not sure wahat to do about the ocaml package. @bcdarwin could you take a look?

Ocaml-torch hasn't been updated in a year or so, so not surprising it would eventually break. Happy for you to mark broken for now.

**Comment by npatsakula** [comment]
Hello @pinpox!

I've tested new version with hash from PR on `x86_64` (linux + CPU), so maybe we can mark checkbox as completed:
https://github.com/NixOS/nixpkgs/pull/302184

Also I need to mention that 2.2.2 has support of `aarch64` Darwin and it would be great to see it.

@junjihashimoto, it seems that I need to close my PR as a duplicate.

File: pkgs/development/libraries/science/math/libtorch/binary-hashes.nix [hunk 1]
```diff
@@ -1,19 +1,19 @@
-version : builtins.getAttr version {
-  "2.0.0" = {
+version: builtins.getAttr version {
+ "2.2.2" = {
```
Comment (npatsakula) [inline pkgs/development/libraries/science/math/libtorch/binary-hashes.nix:2]:
```suggestion
 "2.2.2" = {
     aarch64-darwin-cpu = {
      name = "libtorch-macos-arm64-2.2.2.zip";
      url = "https://download.pytorch.org/libtorch/cpu/libtorch-macos-arm64-2.2.2.zip";
      hash = "sha256-zXviMxGW0wpJDIbb++qaJLL8OwJPuT38KqRRXkEFGeA=";
    };
```

Hello @pinpox! `aarch64` support snippet ^

**Comment by pinpox** [comment]
`aarch64` added (thanks @npatsakula !). Let me know if something else is missing.

File: pkgs/development/libraries/science/math/libtorch/prefetch.sh [hunk 2]
```diff
@@ -6,10 +6,10 @@ set -eou pipefail
 version=$1
 
 bucket="https://download.pytorch.org/libtorch"
-CUDA_VERSION=cu116
+CUDA_VERSION=cu121
 
 url_and_key_list=(
-  "x86_64-darwin-cpu $bucket/cpu/libtorch-macos-${version}.zip libtorch-macos-${version}.zip"
+  "x86_64-darwin-cpu $bucket/cpu/libtorch-macos-x86_64-${version}.zip libtorch-macos-x86_64-${version}.zip"
```
Comment (mrozycki) [inline pkgs/development/libraries/science/math/libtorch/prefetch.sh:12]:
```suggestion
  "aarch64-darwin-cpu $bucket/cpu/libtorch-macos-arm64-${version}.zip libtorch-macos-arm64-${version}.zip"
  "x86_64-darwin-cpu $bucket/cpu/libtorch-macos-x86_64-${version}.zip libtorch-macos-x86_64-${version}.zip"
```

You should probably also add aarch64 to the prefetch script

**Review by junjihashimoto — APPROVED** [review]
LGTM!

**Review by msanft — APPROVED** [review]

**Comment by pbsds** [comment]
Please squash the commits, and could you look into the aarch64-darwin issue?

**Comment by pinpox** [comment]
> Please squash the commits, and could you look into the aarch64-darwin issue?

fixed and squashed

**Review by junjihashimoto — APPROVED** [review]
Thx!

**Comment by pbsds** [comment]
aarch64-darwin is still failing, did you forget to push the change?

**Comment by npatsakula** [comment]
> aarch64-darwin is still failing, did you forget to push the change?

Hello @pbsds! My bad, looks like I need your help:

1. I don't understand this [error message](https://github.com/NixOS/nixpkgs/pull/302427/checks?check_run_id=24033577077) and I am unable to associate it to particular package test.
2. I can't reproduce this test on my `aarch64-darwin` machine, It compiles without errors (I'm using this package from flake).

**Comment by pbsds** [comment]
I'm not familiar with nor have access to darwin, but as I understand the darwin build sandbox is lacking. Perhaps it passes on your machine since the dyld check is able to match with a dylib on your system that the ofborg builders don't have installed?

Comparing the fixupPhase in the full logs: https://logs.ofborg.org/?attempt_id=e65c3bce-5f6e-450f-9be7-f390bfbb6790&key=nixos%2Fnixpkgs.302427

x86_64-darwin:

```
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libtorch_python.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libbackend_with_compiler.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libtorch.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libtorch_global_deps.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libiomp5.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libtorch_cpu.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libjitbackend_test.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libc10.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libtorchbind_test.dylib: fixing dylib
/nix/store/h7rmvcdf3vlizf0nzjadxy18vh60ngin-libtorch-2.2.2/lib/libshm.dylib: fixing dylib
```

aarch64-darwin:

```
/nix/store/2cfbhxddvg39fc6ldqk0kk7qs7vgrigz-libtorch-2.2.2/lib/libtorch_python.dylib: fixing dylib
/nix/store/2cfbhxddvg39fc6ldqk0kk7qs7vgrigz-libtorch-2.2.2/lib/libtorch.dylib: fixing dylib
/nix/store/2cfbhxddvg39fc6ldqk0kk7qs7vgrigz-libtorch-2.2.2/lib/libtorch_global_deps.dylib: fixing dylib
/nix/store/2cfbhxddvg39fc6ldqk0kk7qs7vgrigz-libtorch-2.2.2/lib/libtorch_cpu.dylib: fixing dylib
/nix/store/2cfbhxddvg39fc6ldqk0kk7qs7vgrigz-libtorch-2.2.2/lib/libc10.dylib: fixing dylib
/nix/store/2cfbhxddvg39fc6ldqk0kk7qs7vgrigz-libtorch-2.2.2/lib/libshm.dylib: fixing dylib
```

Some seem to be missing

**Comment by junjihashimoto** [comment]
https://github.com/NixOS/nixpkgs/blob/dd25cfdbaf5e12c320b29064d073d70263923099/pkgs/development/libraries/science/math/libtorch/bin.nix#L24

Could you update the code to use `stdenv.isDarwin`?

**Review by junjihashimoto — APPROVED** [review]
Thx!

**Comment by pbsds** [comment]
builder feilure... I'd love to verify the darwin fix

@ofborg build libtorch-bin

**Comment by junjihashimoto** [comment]
@pinpox @pbsds 

Probably `platforms` should be updated like this PR.

```nix
    platforms = with platforms; [
      "aarch64-darwin"
      "x86_64-darwin" "x86_64-linux"
    ];
```

https://github.com/pinpox/nixpkgs/pull/2/files#diff-3a9f32ef2e590562647c6c78cd8af994c4ef85795f4ef1744e5d43069bdc3155R102-R105

**Comment by pbsds** [comment]
It worked before the push, and now it failed again. 
The failure seems to be a lack of libomp.dylib, which i assume is builder dependent.
Pytorch darwin assumes libomp (https://github.com/pytorch/pytorch/issues/20030) which should be provided by llvmPackages.openmp. If the binaries are patched with this we should get a more hermetic runtime closure

**Comment by wegank** [comment]
Superseded by #308328.

**Pull request closed without merging**